### PR TITLE
Added floating recipient bar.

### PR DESCRIPTION
**Pull request by vishwesh3**
Fix:#385

**Summary of changes**

Screenshots or a brief video showing the change in action:
ezgif com-video-to-gif 9

Please make sure these boxes are checked before submitting your pull request - thanks! [Guide](./PULL_REQUEST_GUIDE.md)

- [x] Code is formatted.

- [x] Run the lint tests with `./gradlew lintDebug` and make sure BUILD is SUCCESSFUL

- [x] If new feature is implemeted then it is compatible with Night theme too.

- [x] Commit messages are well-written.

### CODE REVIEW

**Comment by vishwesh3** [comment]
Update

Now Android message list resembles with web message list.

- Floating recipient bar.
- Recipient bar contains date.
- Message list have date separator between messages.
- Hide avatar and sender name if send by same sender.
- Show `edited` tag, star status and time in left for the second message onward (messages send by same user)
- Hide timestamp of messages send in small interval by same sender.
- Message tile contains only time (date in recipient bar and date separator like in web)

**Screenshots**

screenshot_20170314-032110
screenshot_20170314-032118
screenshot_20170314-032130
screenshot_20170314-032213
screenshot_20170314-032332

**Comment by vishwesh3** [comment]
resolved merge conflicts.

**Comment by kunall17** [comment]
@vishwesh3 this looks awesome, just noticed one problem if you scroll towards older messages (going up) then the toolbar is made visible, and the sticky header gets hidden, 

Any solutions to it?

**Comment by vishwesh3** [comment]
@kunall17 One possible solution: - now sticky headers floats below toolbar if it is shown. 

ezgif com-video-to-gif 27

**Comment by kunall17** [comment]
@vishwesh3 

Faced problems like these
This can be fixed! 
[screenshot_20170419-185828](https://cloud.githubusercontent.com/assets/12700799/25182471/4d38b962-2532-11e7-98de-5196f91edf32.png)

**Comment by vishwesh3** [comment]
@kunall17 Updated :+1: 

This is how it look's now.
ezgif com-optimize

File: app/src/main/java/com/zulip/android/viewholders/stickyheaders/interfaces/RetrieveHeaderView.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+import android.support.v7.widget.RecyclerView;
+import android.view.ViewGroup;
+
+public interface RetrieveHeaderView {
+
+    int DEFAULT_VIEW_TYPE = -1;
+
+    RecyclerView.ViewHolder getViewHolderForPosition(int headerPositionToShow);
+
+    class RecyclerRetrieveHeaderView implements RetrieveHeaderView {
```
Comment (kunall17) [inline app/src/main/java/com/zulip/android/viewholders/stickyheaders/interfaces/RetrieveHeaderView.java]: This file class structure can be improved!

**Comment by kunall17** [comment]
Also a little layout issue, the timestamp is not vertically centered! 

device-2017-04-22-001751

File: app/src/main/java/com/zulip/android/viewholders/stickyheaders/interfaces/RetrieveHeaderView.java (hunk 1 above)
Comment (vishwesh3) [inline app/src/main/java/com/zulip/android/viewholders/stickyheaders/interfaces/RetrieveHeaderView.java]: Done :+1:

**Comment by kunall17** [comment]
Cool, @vishwesh3 there are multiple yesterday, today tags coming up is that intentional or a bug? 

(Even if its a bug, I don't have much problems with it) would like to know thought's of other folks @Sam1301 ?

**Comment by Sam1301** [comment]
> (Even if its a bug, I don't have much problems with it) would like to know thought's of other folks @Sam1301 ?

The floating headers look really nice; I am in favour of merging this. Though I feel like the timestamps are *too small* to read.

Also we might want to wrap the header topics to max 2 lines since on small screen devices, it looks something like this 
whatsapp image 2017-04-22 at 18 54 02

**Comment by kunall17** [comment]
That looks awful, the main problem with this one the stream name is also really long hence even for two lines we won't get much space and end up loosing characters. 

How about we go for ellipsis for the topic textview?

**Comment by vishwesh3** [comment]
if we go with `ellipsize="end" ` and `maxLines="1"` or `maxLines="2"`, then there should be some place where user can find whole topic name. (as in topic narrow it is ellipsize)
cb95a3ed-7c2c-4db1-9935-fc25e89efe65

@Sam1301 What is the screen size of the [above image](https://github.com/zulip/zulip-android/pull/437#issuecomment-296373509)?

**Comment by Sam1301** [comment]
@vishwesh3 ldpi (320 X 426 in dp)

**Comment by vishwesh3** [comment]
One solution :- we can limit stream, topic name to `ellipsize="end"` and `maxLines="1"` or `maxLines="2"` . And on long press (stream or topic) we can show full name in the header itself or in a toast.

Thoughts?

**Comment by kunall17** [comment]
Yeah, i think maxlines = 1 would be fine (2 would be too much packed up) 
We can use the click there in the toolbar as noticing long press in the toolbar is a little difficult and no one would expect that, on click there can be more possibility?

**Comment by vishwesh3** [comment]
Yes, we can also make toolbar clickable.

I was thinking, long press on message header.

**Comment by vishwesh3** [comment]
Like this
ezgif com-video-to-gif 29

**Comment by divs4debu** [comment]
I think that would be too much of user interaction on message header.

**Comment by divs4debu** [comment]
What u can do is that when the header reaches the top it can show full topic and that might look good too.

**Comment by kunall17** [comment]
@vishwesh3 This interaction seems nice, but what about the topic header as Saumya mentioned, that would create problem, but otherwise this seems good.

I would like to wrap up this PR with this interaction, so rebase and we'll do a release! :)
And if you're covering the toolbar click handler too do them in different commits! 

@divs4debu If we do something like that, while scrolling the header will be constantly getting big and small which would not look good!

**Comment by vishwesh3** [comment]
rebased

**Comment by kunall17** [comment]
Awesome work, merged thanks @vishwesh3 👍

**Pull request merged**